sycl/source/detail/graph_impl.cpp

```diff
@@ -913,6 +912,149 @@ exec_graph_impl::enqueue(const std::shared_ptr<sycl::detail::queue_impl> &Queue,
   return QueueEvent;
 }
 
+void exec_graph_impl::duplicateNodes() {
+  // Map of original modifiable nodes (keys) to new duplicated nodes (values)
+  std::map<std::shared_ptr<node_impl>, std::shared_ptr<node_impl>> NodesMap;
+
+  std::vector<std::shared_ptr<node_impl>> ModifiableNodes =
+      MGraphImpl->MNodeStorage;
+  std::deque<std::shared_ptr<node_impl>> NewNodes;
+
+  for (int i = ModifiableNodes.size() - 1; i >= 0; i--) {
+    auto OriginalNode = ModifiableNodes[i];
+    std::shared_ptr<node_impl> NodeCopy =
+        std::make_shared<node_impl>(*OriginalNode);
+
+    // Clear edges between nodes so that we can replace with new ones
+    NodeCopy->MSuccessors.clear();
+    NodeCopy->MPredecessors.clear();
+    // Push the new node to the front of the stack
+    NewNodes.push_front(NodeCopy);
+    // Associate the new node with the old one for updating edges
+    NodesMap.insert({OriginalNode, NodeCopy});
+  }
+
+  // Now that all nodes have been copied rebuild edges on new nodes. This must
+  // be done as a separate step since successors may be out of order.
+  for (size_t i = 0; i < ModifiableNodes.size(); i++) {
+    auto OriginalNode = ModifiableNodes[i];
+    auto NodeCopy = NewNodes[i];
+    // Look through all the original node successors, find their copies and
+    // register those as successors with the current copied node
+    for (auto &NextNode : OriginalNode->MSuccessors) {
+      auto Successor = NodesMap.at(NextNode.lock());
+      NodeCopy->registerSuccessor(Successor, NodeCopy);
+    }
+  }
+
+  // Subgraph nodes need special handling, we extract all subgraph nodes and
+  // merge them
+  // into the main node list
+
+  for (auto NewNodeIt = NewNodes.end(); NewNodeIt != NewNodes.begin();) {
+    --NewNodeIt;
+    auto NewNode = *NewNodeIt;
+    if (NewNode->MNodeType == node_type::subgraph) {
```

Comment:
Preference for doing an early continue to avoid the extra indentation for the majority of the for loop.
```cpp
    if (NewNode->MNodeType != node_type::subgraph) {
      continue;
    }
```